core/src/Cuda/Kokkos_Cuda_KernelLaunch.hpp

```diff
@@ -205,15 +207,53 @@ modify_launch_configuration_if_desired_occupancy_is_specified(
 
   if (dynamic_shmem > shmem) {
     shmem        = dynamic_shmem;
-    prefer_shmem = false;
+    prefer_shmem = 0;
   }
 }
 
 template <class Policy>
 std::enable_if_t<!Policy::experimental_contains_desired_occupancy>
 modify_launch_configuration_if_desired_occupancy_is_specified(
-    Policy const&, cudaDeviceProp const&, cudaFuncAttributes const&,
-    dim3 const& /*block*/, int& /*shmem*/, bool& /*prefer_shmem*/) {}
+    Policy const&, cudaDeviceProp const& properties,
+    cudaFuncAttributes const& attributes, dim3 const& block, int& shmem,
+    int& prefer_shmem) {
+  // prefer_shmem = 0 - cudaFuncCachePreferL1
+  // prefer_shmem = 1 - cudaFuncCachePreferEqual
+  // prefer_shmem = 2 - cudaFuncCachePreferShared
+
+  // If prefer_shmem is already set to 0, return.
+  if (prefer_shmem == 0) return;
+
+  // Calculate maximum number of blocks that can simultaneously run on a SM
+  // based on the block size requested.
+  int const block_size   = block.x * block.y * block.z;
+  int max_blocks_threads = properties.maxThreadsPerMultiProcessor / block_size;
+
+  // Calculate the maximum number of blocks that can simultaneously run based on
+  // the number of registers.
+  int const regs_per_sm     = properties.regsPerMultiprocessor;
+  int const regs_per_thread = attributes.numRegs;
+  int const max_blocks_regs = regs_per_sm / (regs_per_thread * block_size);
+
+  // Set active_blocks to be the minimum of the two.
+  int const active_blocks = std::min({max_blocks_threads, max_blocks_regs});
+
+  // Returns approximately half of the configurable cache size.
+  size_t const shmem_per_sm_prefer_equal =
+      get_shmem_per_sm_prefer_equal(properties);
+  size_t const static_shmem = attributes.sharedSizeBytes;
+
+  // Caclulate how much shmem will be assigned per active block.
+  int const dynamic_shmem =
+      shmem_per_sm_prefer_equal / active_blocks - static_shmem;
+
+  // If dynamic_shmem is greater than requested shmem, set
+  // cudaFuncCachePreferEqual, else set cudaFuncCachePreferShared.
```

Comment:
Why no check whether the L1 preference is already enough shared memory?

Reply:
My understanding was that the preference for L1 is already being calculated before reaching this point. If that's not the case, I can add it here.

Reply:
Added a preference for L1.